src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,243 @@
+            out int tokenTypeOut,
+            out int tokenModifiersOut)
+        {
+            // Each semantic token is represented in LSP by five numbers:
+            //     1. Token line number delta, relative to the previous token
+            //     2. Token start character delta, relative to the previous token
+            //     3. Token length
+            //     4. Token type (index) - looked up in SemanticTokensLegend.tokenTypes
+            //     5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+
+            var classifiedSpan = classifiedSpans[currentClassifiedSpanIndex];
+            var linePosition = lines.GetLinePositionSpan(classifiedSpan.TextSpan).Start;
+            var lineNumber = linePosition.Line;
+            var startCharacter = linePosition.Character;
+
+            // 1. Token line number delta, relative to the previous token
+            var deltaLine = lineNumber - lastLineNumber;
+            Contract.ThrowIfTrue(deltaLine < 0);
+
+            // 2. Token start character delta, relative to the previous token
+            // (Relative to 0 or the previous token’s start if they're on the same line)
+            var startCharacterDelta = startCharacter;
+            if (lastLineNumber == lineNumber)
+            {
+                startCharacterDelta = startCharacter - lastStartCharacter;
+            }
+
+            // 3. Token length
+            var tokenLength = classifiedSpan.TextSpan.Length;
+
+            // We currently only have one modifier (static). The logic below will need to change in the future if other
+            // modifiers are added in the future.
+            var modifierBits = TokenModifiers.None;
+            var tokenTypeIndex = 0;
+            var originalTextSpan = classifiedSpan.TextSpan;
+
+            // Classified spans with the same text span should be combined into one token.
+            while (classifiedSpans[currentClassifiedSpanIndex].TextSpan == originalTextSpan)
+            {
+                var classificationType = classifiedSpans[currentClassifiedSpanIndex].ClassificationType;
+                if (classificationType != ClassificationTypeNames.StaticSymbol)
+                {
+                    // 4. Token type - looked up in SemanticTokensLegend.tokenTypes (language server defined mapping
+                    // from integer to LSP token types).
+                    tokenTypeIndex = GetTokenTypeIndex(classificationType, tokenTypesToIndex);
+                }
+                else
+                {
+                    // 5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+                    modifierBits = TokenModifiers.Static;
+                }
+
+                if (currentClassifiedSpanIndex + 1 >= classifiedSpans.Length || classifiedSpans[currentClassifiedSpanIndex + 1].TextSpan != originalTextSpan)
+                {
+                    break;
+                }
```

Comment:
```suggestion
```

Reply:
Hmm.. I looked into this, but I don't know if the logic is equivalent, since we only want to increment `currentClassifiedSpanIndex` in certain scenarios.